### ISSUE TITLE
Compliations of issues/questions from my experience

### ISSUE THREAD

**Issue by Kenny-Hui** [leezer3/OpenBVE#772]
## Possible Bugs (Minor)
~**1.** This panel does not have a theme set when initially opening the menu. (Both in Windows & Linux)
This could be fixed by switching to any other tab, then returning to the game selection menu.~
![image](https://user-images.githubusercontent.com/28094366/161421884-0970abc7-3ca4-4837-9835-5380da3087ef.png)

**2.** On Linux, in-game UI fonts often seems to clip through. (nvidia-driver-470)
Though I assume it just has something to do with the Graphics driver?
![image](https://user-images.githubusercontent.com/28094366/161422142-4cc7d99e-79b8-4f55-aea5-777275d0c6d8.png)

## Possible Impactful Bugs (Resolved)
~**1.** After the door width fix of #770, I can't get the train to power anymore. (Whether using a custom plugin or the default ATS)
Reproducible with the same train in #770.~

~**2.** After a car rolls over/flipped, the loop.wav will no longer play on that car.
However after Jumping Stations, the loop.wav remains silence.~

~**3.** The driver position seems to be moved forward after using CarXmlConverter
**Train.dat:**
0, 2.385, -1.825 (0, 2385, -1825)
**Train.xml:**
0, 2.385, 8.425~

~**4.** #767~

## Questions (Resolved)
~**1.** The train.xml allows you to set a custom Notch Description.
The train has 8 Power Notch and 7 Brake Notch, and I added the following to the xml file:~
```
<NotchDescriptions>
	<Power>10deg,15deg,20deg,25deg,30deg,35deg,40deg,45deg</Power>
	<Brake>10deg,15deg,20deg,25deg,30deg,35deg,40deg</Brake>
</NotchDescriptions>
```
~In-Game I am only seeing the Neutral Notch/EMG displaying `10deg,15deg,20deg,25deg,30deg,35deg,40deg,45deg`.
![image](https://user-images.githubusercontent.com/28094366/161423934-5981ea42-9411-4252-b366-3ac10a699fa8.png)
Am I doing something wrong?~

~**2.** I have not tried this for a while now, but a few months ago when I set the Horn Sound to a loopable one (one start sound, one loop sound), it seems there will always be a pause after playing the start sound, before start looping.
Is there a way to make it play smoothly?~

~**3.** When I reload a extension.cfg along with a train.xml in Object Viewer, the following error will pop up:
`The train plugin LRVSystem.dll failed to load for the following reason: An item with the same key has already been added.`
Is there any way to prevent such error from happening, and why is the train plugin loaded in the ObjectViewer?~

### Related information 

* Operating system: Windows 10 21H2 / Linux Mint 20.3
* Method of control: Keyboard
* OpenBVE Version: 2022-04-02 (Nightly) on both OS
* Using Mono to execute OpenBVE on Linux

**Comment by leezer3** [comment]
Sorted some of the simple ones.

Am looking at the others, may take a while mind....

**Comment by ginga81** [comment]
In my environment, mainmenu's buttons are blue.
This is by OpenBVE-2022-04-04.zip.
And continuing https://github.com/leezer3/OpenBVE/issues/721
![blue-mainmenu](https://user-images.githubusercontent.com/22734699/161560069-d3b1cc4f-d7ca-4b02-92d0-5409873c374b.png)

**Comment by leezer3** [comment]
Plugins in Object Viewer:
I can't reproduce this at the minute. I suspect it's most likely the plugin in question not liking multiple instances being loaded, but that's only a suspicion.....

The reason for doing it this way, is mainly because everything now loads through the single unified interface. We have the existing train details window, and I suppose an end-aim will be to have plugin simulation inside the viewers, so that some of the more complex setups can be tested properly.
It shouldn't do anything (much) at the minute though, the plugin update method is never called.....

**Comment by Kenny-Hui** [comment]
> Plugins in Object Viewer: I can't reproduce this at the minute. I suspect it's most likely the plugin in question not liking multiple instances being loaded, but that's only a suspicion.....
> 
> The reason for doing it this way, is mainly because everything now loads through the single unified interface. We have the existing train details window, and I suppose an end-aim will be to have plugin simulation inside the viewers, so that some of the more complex setups can be tested properly. It shouldn't do anything (much) at the minute though, the plugin update method is never called.....

Nevermind, the plugin is trying to register a list of translation twice, but since it doesn't expect the Load() method more than once it errors out.

Speaking of which, is there a good alternative to Windows Forms for like a Config Panel?
![image](https://user-images.githubusercontent.com/28094366/161684970-f53bf6dc-f380-4759-9ed0-3dcec6df6add.png)
I am afraid there's not enough virtual key to bind all of these.
The forms works quite well on Windows, but it seems rather unstable under Mono

**Comment by Kenny-Hui** [comment]
![image](https://user-images.githubusercontent.com/28094366/161724982-8fa7eb96-5d73-49ae-9b79-5ea4efdb4989.png)
Also you will not be able to get rid of this error if it can't switch to full screen during gameplay.
It will continue to pop up this message every frame, until you forcefully terminate the program

**Comment by leezer3** [comment]
Fixed the error; the error box was blocking the key up event...

Winforms config window:
This is a mess, there isn't really a good solution.
Essentially, due to some assumptions in Mono code, a WinForm must control the event loop. You'll find that if you start doing anything complex with it, you can freeze the sim (IIRC try something like a simple 10s blocking timer)

The only decent way around it is to move the entire thing out of process, for which you'd have to implement a communication stack somehow. The Windows 32-bit plugin proxy does this via named pipes, but last I checked this wasn't too happy on older Mono, so you'd have to do something like an embedded webserver or something :/

**Comment by leezer3** [comment]
Notch descriptions seems to have been using a semi-colon as the separator char.
I've changed this so it accepts either a comma or a semi-colon.

**Comment by Kenny-Hui** [comment]
Just checked on the horn, using sound.xml to define the following:
```
<Secondary>
	<Start>
		<FileName>Sound\System\Horn1.wav</FileName>
		<Position>0, 0, 10</Position>
		<Radius>30.0</Radius>
	</Start>
	<Loop>
		<FileName>Sound\System\Horn2.wav</FileName>
		<Position>0, 0, 10</Position>
		<Radius>30.0</Radius>
	</Loop>
	<End>
		<FileName>Sound\System\Horn3.wav</FileName>
		<Position>0, 0, 10</Position>
		<Radius>30.0</Radius>
	</End>
</Secondary>
```
Doesn't seems to have any difference than simply a normal horn (Only the <loop> section played once)
Using sound.cfg works properly, but there's still a small pause between start & loop (Despite there's no silence on the original wav file)

https://user-images.githubusercontent.com/28094366/161889469-ad0910ee-e7b0-4ac5-ab13-ecb5617e064e.mp4

**Comment by adfriz** [comment]
@Kenny-Hui are you have tried make the sound overlapping ? Maybe it can solve that pause.

Let say your original sound files is 8 seconds long sound but you cut into 3 part sound,
Start.wav is 0-2
Loop.wav is 3-6
End is 7-8

Maybe you need overlapping the sound by 1 second, so the sound like this:
Start.wav is 0-2
Loop.wav is 2-6
End.wav is 6-8

I havent try this method, but maybe this will works.

**Comment by Kenny-Hui** [comment]
> I doubt maybe currently... channel 1:|------start----||------continuing----||------end----|
> 
> My suggestion... channel 1:|------start----| channel 2:                    |------continuing----| channel 3:                                                  |------end----|

Yea that would probably work better with solving the poping sound

The main problem I think is OpenBVE will just wait for the sound to end, then play another sound.
(But there will always a small gap after the start ended, and before the loop plays)

**Comment by adfriz** [comment]
@Kenny-Hui yeah you right changing current behavior would make more mess and make some addons creator wasting their time to relearn the new behavior.

**Comment by ginga81** [comment]
Pop sound, that is also happen at F81_tec200's Series-E4.
https://youtu.be/-bjdICkOPD0

https://user-images.githubusercontent.com/22734699/162007905-d98c0124-1cc1-41d6-a4d6-17f4a2d4cdcf.mp4

**Comment by adfriz** [comment]
@ginga81 thats also my second though, either do my suggest above or stack the sound like your say.

**Comment by adfriz** [comment]
@ginga81 my addons sound like that too, i dont know why it seems the pause very noticeable when transitioning from start sound to loop sound but usually from loop to end is very smooth.
It seems from start to loop took longer to process but from loop to end is pretty quick to process.

**Comment by ginga81** [comment]
@adfriz 
Such as this confusion has been already happening at setemmisivecolor.
loadtexture, daytimeimage, daytimeimage
This command has been long time for use easily self lighting.
But, at the GL3.0, we have to use SetEmissiveColor.
This change is very big.
But we are overcoming problems.
So I think you don't have to worry about it.

**Comment by leezer3** [comment]
The start sound issue will be frame timing related.
First, we've got to detect that the sound *has* ended (IIRC via an OpenAL call, rather than keeping track ourselves), then enqueue and play the new sound.
This will take two frames to do so. Depending on your FPS you may see a slight delay. Keeping track of the duration ourselves would allow enqueing the sound earlier, but still leaves you at the mercy of frame timing.

The end sound on the other hand will issue a stop command, followed by a play for the new sound in the same frame, and hence appears instantaneous. (As the loop continues to play whilst the processing occurs)

TLDR:
It's technically a massive pain...
No good easy solution off the top of my head.
The average length of horn sounds makes this even trickier.

**Comment by Kenny-Hui** [comment]
Hm but I still can't get a loopable horn working with sound.xml (sound.cfg works) using the above code

**Comment by leezer3** [comment]
Not looked at that, will try and do so.

**Comment by ginga81** [comment]
Relative https://github.com/leezer3/OpenBVE/issues/778.
I tested the newest daily build.
At my monodevelop, nothing the errors.
And Mainmenu's grayout is fixed at the newest daily build.
Thank you very much for fix!
![Screenshot from 2022-04-12 21-25-07](https://user-images.githubusercontent.com/22734699/162962441-4b226745-ef32-4d3b-9c3d-fbc478c183b2.png)
